### PR TITLE
Fixes for clang

### PR DESCRIPTION
**Pull request by niboshi**
Checked without CUDA backend with:
* clang version 5.0.1-4
* clang version 6.0.0-1ubuntu2

### CODE REVIEW

File: chainerx_cc/chainerx/routines/manipulation_test.cc [hunk 1]
```diff
@@ -839,7 +839,9 @@ TEST_THREAD_SAFE_P(ManipulationTest, SplitDefaultAxesIndices) {
     Array e1 = testing::BuildArray({1, 4}).WithData<int32_t>({0, 1, 2, 3});
     Array e2 = testing::BuildArray({1, 4}).WithData<int32_t>({4, 5, 6, 7});
 
-    Run([&]() { testing::CheckForward([](const std::vector<Array>& xs) { return Split(xs[0], {2}); }, {a}, {e1, e2}); });
+    Run([&]() {
+        testing::CheckForward([](const std::vector<Array>& xs) { return Split(xs[0], std::vector<int64_t>{1}); }, {a}, {e1, e2});
```
Comment (asi1024) [inline chainerx_cc/chainerx/routines/manipulation_test.cc]: Typo.
```suggestion
        testing::CheckForward([](const std::vector<Array>& xs) { return Split(xs[0], std::vector<int64_t>{2}); }, {a}, {e1, e2});
```

Comment (niboshi) [reply chainerx_cc/chainerx/routines/manipulation_test.cc]: This fix is intentional.
Please check the shapes.

**Comment by hvy** [comment]
Have we decided which versions of clang to support?

**Comment by hvy** [comment]
Note that it still fails to build with Apple clang. This conversion (to the return type defined by the signature) seems invalid. Could you take a look? https://github.com/chainer/chainer/blob/7f539ef53c3b2ff22e4203ea7f08250889e29545/chainerx_cc/chainerx/native/native_device/memory.cc#L18

**Comment by hvy** [comment]
Sorry forgot this comment. Could you address it? It LGTM otherwise. https://github.com/chainer/chainer/pull/5744#pullrequestreview-181162212

File: chainerx_cc/CMakeLists.txt [hunk 2]
```diff
@@ -108,6 +108,21 @@ if("${CMAKE_CXX_COMPILER_ID}" MATCHES "GNU|Clang|Intel")
         set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -fsanitize=thread")
     endif()
 endif()
+
+# Compiler-specific workarounds
+
```
[the remark this answers is not in the document]
Comment (hvy) [reply chainerx_cc/CMakeLists.txt]: Discussed offline to leave it as is since we're expecting other workarounds to be placed here in the future.

**Comment by hvy** [comment]
LGTM

**Pull request merged**